modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/KeyCriterion.java

```diff
@@ -118,45 +222,113 @@ public CollectionCriterion(Collection<Key> keys) {
         }
 
         /** {@inheritDoc} */
-        @Override public IgniteBiTuple<Key, Key> toRange() {
-            return new IgniteBiTuple<>(Collections.min(keys), Collections.max(keys));
+        @Override public boolean contains(Key key) {
+            return keys.contains(key);
         }
 
         /** {@inheritDoc} */
-        @Override public boolean contains(Key key) {
-            return keys.contains(key);
+        @Override protected Optional<KeyCriterion> doUnion(KeyCriterion keyCriterion) {
+            var newKeys = new ArrayList<>(keys);
```

Comment:
Hmm, why not to use Set? Am i right that in current solution it's possible to duplicate keys?